### PR TITLE
`Headers.fromEntries`: using `CaseInsensitiveMap.fromEntries`

### PR DESCRIPTION
**Pull request by gmpassos**
* `Headers`:
  * Added the constructor `Headers.fromEntries`.
  * Added the internal constructor `_fromEntries`, which uses `CaseInsensitiveMap.fromEntries` for optimization.
  * Renamed the internal constructor `_` to `_from`, redirecting it to `this._fromEntries`.

---

- [✅ ] I’ve reviewed the contributor guide and applied the relevant portions to this PR.

### CODE REVIEW

**Comment by gmpassos** [comment]
FYI: @kevmoo

**Review by kevmoo — APPROVED** [review]
I made some tweaks.

**Comment by kevmoo** [comment]
@gmpassos – and some more tweaks 😄

File: pkgs/shelf/lib/src/headers.dart [hunk 1]
```diff
@@ -23,15 +23,24 @@ class Headers extends UnmodifiableMapView<String, List<String>> {
       return _emptyHeaders;
     } else if (values is Headers) {
       return values;
+    } else {
+      return Headers._(values.entries);
+    }
+  }
+
+  factory Headers.fromEntries(
+      Iterable<MapEntry<String, List<String>>>? values) {
```
Comment (gmpassos) [inline pkgs/shelf/lib/src/headers.dart:32]: Maybe rename to `entries`?

Comment (kevmoo) [reply pkgs/shelf/lib/src/headers.dart:32]: done

Comment (kevmoo) [reply pkgs/shelf/lib/src/headers.dart]: yep

**Comment by kevmoo** [comment]
@gmpassos – do you have benchmarks showing this makes things notably faster?

**Comment by gmpassos** [comment]
> @gmpassos – we need to roll this through our internal "stuff" before we publish. To make sure nothing breaks. I'll let you know when publish happens!

Thanks for the quick response and attention. I will wait for the publication...

**Comment by kevmoo** [comment]
Looks like @devoncarew landed in the Dart SDK 50 minutes ago. We'll have to wait a few hours to make sure that rolls into our internal...things...before we publish

**Comment by gmpassos** [comment]
> @gmpassos – do you have benchmarks showing this makes things notably faster?

In my benchmark, I get a 1-2% improvement, but note that you need to have real-world headers (real browsers send many more headers than simple HTTP clients). Additionally, you need a heavy load, as this also affects the GC. This is more significant for slower/smaller servers. On a medium-sized server without low traffic, this won't have much impact.

I'm also looking for a way to avoid computation of all canonicalized keys just to build the headers. This will significantly improve performance, or we could use a specialized cache of canonicalized keys, but this will be a future optimization.

**Comment by kevmoo** [comment]
shelf published w/ this! @gmpassos !

**Pull request merged**